refactor(core): add generic utilities for resolving value-or-function patterns, replace specialized `resolveStaleTime` and `resolveEnabled`

braden-w opened the pull request:
This PR introduces `isFunctionVariant()` and `resolveValueOrFunction()` utilities to replace repetitive `typeof === 'function'` checks throughout the codebase, consolidating the common "value or function that computes value" pattern.

## Problem and Solution

The codebase had scattered implementations of the same pattern across multiple files:

```typescript
// Repetitive pattern everywhere
const data = typeof options.initialData === 'function' 
  ? options.initialData() 
  : options.initialData

const delay = typeof retryDelay === 'function'
  ? retryDelay(failureCount, error) 
  : retryDelay
```

This led to code duplication, inconsistency, and maintenance overhead. We also had specialized versions like `resolveStaleTime()` and `resolveEnabled()` that could be generalized.

The new utilities provide a clean, generic solution:

```typescript
// Clean and intention-revealing
const data = resolveValueOrFunction(options.initialData)
const delay = resolveValueOrFunction(retryDelay, failureCount, error)
```

While we could inline `typeof value === 'function'` checks, TypeScript's type narrowing doesn't work properly with generic types in complex expressions. The `isFunctionVariant()` type guard provides proper type narrowing that allows `resolveValueOrFunction()` to safely call the function variant. Without it, TypeScript throws errors because it can't guarantee the type safety across the generic boundary.

Both utilities support zero-argument functions (`() => T`) and functions with parameters (`(...args) => T`), making them applicable to all value-or-function patterns in the codebase.


Updated implementations in `query.ts`, `queryObserver.ts`, and `retryer.ts` for handling `initialData`, `retryDelay`, `refetchInterval`, and `notifyOnChangeProps`. These utilities can replace existing `resolveStaleTime()` and `resolveEnabled()` functions in future iterations.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Placeholder data can now be provided as either a direct value or a function for more flexible initial UI states.

* **Refactor**
  * Consolidated option handling into a single, consistent resolution approach used for enabled checks, stale timing, initial/placeholder data, retry delays, and observer notifications.

* **Chores**
  * Removed legacy per-option helpers and some exported helper types; introduced a unified option-resolver utility and consolidated related logic.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

braden-w committed:
refactor(core): add generic utilities for resolving value-or-function patterns, replace specialized `resolveStaleTime` and `resolveEnabled`

This commit introduces `isFunctionVariant()` and `resolveValueOrFunction()` utilities to replace repetitive `typeof === 'function'` checks throughout the codebase, consolidating the common "value or function that computes value" pattern.

braden-w committed:
refactor(types): simplify placeholderData type definition by removing NonFunctionGuard

This simplification is possible due to the introduction of generic runtime
utilities that handle value-or-function resolution. The new `resolveValueOrFunction()` utility handles the distinction between direct values and functions at runtime with proper type safety, eliminating the  need for complex type-level guards.

braden-w committed:
Merge branch 'upstream/main' into refactor/resolveValueOrFunction

braden-w committed:
chore(query): replace resolveStaleTime with resolveValueOrFunction

braden-w committed:
chore(deps): add tsup dependency and fix build

TkDodo committed:
Merge branch 'main' into refactor/resolveValueOrFunction

TkDodo reviewed (commented):
I generally agree with the direction, please have a look at the comments though

braden-w committed:
Merge upstream/main into refactor/resolveValueOrFunction

braden-w committed:
Merge remote-tracking branch 'origin/main' into refactor/resolveValueOrFunction

braden-w committed:
refactor: rename resolveValueOrFunction to resolveOption for consistency

autofix-ci[bot] committed:
ci: apply automated fixes

braden-w committed:
refactor(queryClient): update method call to use resolveOption for improved clarity

braden-w committed:
refactor: replace resolveEnabled with resolveOption for consistency

Remove specialized resolveEnabled function and consolidate functionality
into the generic resolveOption utility. Both functions performed identical
operations but resolveEnabled was type-specific to query enabled state.

- Remove resolveEnabled function from utils.ts
- Update queryObserver.ts to use resolveOption instead
- Remove unused Enabled type import
- Maintain identical functionality and type safety

autofix-ci[bot] committed:
ci: apply automated fixes

TkDodo commented:
it’s been two months, there’s still a failing pipeline and conflicts. please let me know if you intend to finish this PR, because it doesn’t look like it’s close to working. I’ll close it otherwise.

braden-w commented:
Thanks for the ping! Will update it and get back to you by EOD.

braden-w committed:
chore: remove unused dependency tsup from package.json

braden-w commented on packages/query-core/src/utils.ts:
````diff
@@ -80,6 +77,94 @@ export function noop(): void
+ * ```
+ */
+export function resolveValueOrFunction<T, TArgs extends Array<any>>(
+  value: T | ((...args: TArgs) => T),
````
Seems like NonFunctionGuard is no longer in the codebase, so I'll refrain from adding it for now!

braden-w committed:
refactor: simplify functionalUpdate by using resolveOption

braden-w committed:
Merge remote-tracking branch 'upstream/main' into refactor/resolveValueOrFunction

braden-w committed:
fix: correct spelling of 'retryer' in comment

braden-w committed:
chore: pnpm lock

TkDodo committed:
Merge branch 'main' into refactor/resolveValueOrFunction

braden-w committed:
refactor(core): constrain resolveOption to non-function types

Add NonFunction type constraint to resolveOption's generic parameter T to prevent
ambiguity when T itself is a function type. This ensures the value-or-function
pattern remains unambiguous by restricting T to primitives, objects, and other
non-function types.

- Define NonFunction union type covering all non-function types
- Apply constraint to resolveOption<T extends NonFunction, ...>
- Update functionalUpdate to use same constraint
- Remove isFunctionVariant helper (inlined typeof check)
- Simplify resolveOption implementation

This prevents TypeScript errors where function types could be ambiguous in the
T | (() => T) pattern, making the API more type-safe.

braden-w committed:
Merge branch 'main' of https://github.com/TanStack/query into refactor/resolveValueOrFunction

braden-w committed:
docs(core): improve NonFunction type organization and documentation

Move NonFunction type definition above resolveOption function for better logical
flow and add comprehensive JSDoc documentation explaining its purpose in preventing
recursive type ambiguity in value-or-function patterns.

- Add JSDoc for NonFunction type explaining constraint rationale
- Move type definition before its usage for better code organization
- Update resolveOption JSDoc to clarify T is constrained to non-function types
- Improve code readability and developer understanding of type constraints

braden-w committed:
refactor(core): improve resolveOption function with better type safety

- Move NonFunction to generic utility type that excludes function types
- Update Updater type to use NonFunction<TOutput> for better type constraints
- Add function overloads to resolveOption for optional vs required values
- Improve implementation with explicit type assertions and clearer logic
- Remove functionalUpdate function (replaced by resolveOption usage)
- Enhance documentation with better examples and type explanations

braden-w committed:
refactor(query-core): replace functionalUpdate with resolveOption in QueryClient

Replace functionalUpdate usage with resolveOption for consistency with the
new unified value-or-function resolution pattern.

braden-w committed:
docs(core): enhance resolveOption function documentation

Add detailed explanations of NonFunction constraint benefits,
overload behavior for optional vs required values, and improved
type safety guarantees.

autofix-ci[bot] committed:
ci: apply automated fixes

braden-w committed:
Merge remote-tracking branch 'upstream/main' into refactor/resolveValueOrFunction

braden-w committed:
refactor(query-core): consolidate NonFunction type into NonFunctionGuard

Moved NonFunction type from utils.ts to types.ts as NonFunctionGuard with improved documentation. Applied the type guard consistently to placeholderData and updated all references and documentation to use the new name.

braden-w committed:
fix(types): export PlaceholderDataFunction type for better accessibility

braden-w committed:
Merge remote-tracking branch 'upstream/main' into refactor/resolveValueOrFunction

braden-w committed:
refactor(core): use input-driven type inference for resolveOption

Rewrote resolveOption to infer return type from input rather than
requiring NonFunctionGuard<T> constraints. This enables the utility
to work with generic types (TData, TQueryData, etc.) where TypeScript
cannot prove the generic isn't a function.

- Changed signature to use ResolvedValue<TValueOrFn> conditional type
- Converted initialData/initialDataUpdatedAt in query.ts to use resolveOption
- Converted placeholderData in queryObserver.ts to use resolveOption
- Use functional form for setQueryData with generics in streamedQuery.ts

braden-w commented:
Hey @TkDodo! Updated with a rewrite that fixes the type errors.

Renamed to `resolveOption` as you suggested. Regarding `NonFunctionGuard`: I dug into this and found it doesn't work when `T` is a generic type parameter (like `TData`, `TQueryData`) because TypeScript can't evaluate `TData extends Function` at compile time—the conditional type stays unresolved and breaks inference.

The fix uses input-driven type inference instead (same pattern as React's `useState` and Zustand):

```typescript
type ResolvedValue<TValueOrFn> = TValueOrFn extends (...args: Array<any>) => infer R
  ? R
  : TValueOrFn

export function resolveOption<TValueOrFn>(
  valueOrFn: TValueOrFn,
  ...args: Array<unknown>
): ResolvedValue<TValueOrFn>
```

This infers the return type from the actual input type at each call site, sidestepping the generic limitation. For cases where the input involves a generic like `TData`, we use either a type assertion (`as TData | undefined`) or a functional wrapper (`() => result`).

Also unified this with `functionalUpdate`—`setQueryData` now uses `resolveOption` internally. Converted `initialData`, `initialDataUpdatedAt`, and `placeholderData` to use it as well.

Let me know if you'd like any tweaks!

braden-w committed:
fix(core): remove unnecessary type assertion flagged by eslint

TkDodo committed:
Merge branch 'main' into refactor/resolveValueOrFunction

TkDodo commented:
4 months after my last review we now have lots of conflicts. If you want to land this I think it’s easier to file a new PR. Thanks.
